examples/allocation/tinygo/testdata/greet.go

```diff
@@ -63,32 +64,24 @@ func _greeting(ptr, size uint32) (ptrSize uint64) {
 // ptrToString returns a string from WebAssembly compatible numeric types
 // representing its pointer and length.
 func ptrToString(ptr uint32, size uint32) string {
-	// Get a slice view of the underlying bytes in the stream. We use SliceHeader, not StringHeader
-	// as it allows us to fix the capacity to what was allocated.
-	return *(*string)(unsafe.Pointer(&reflect.SliceHeader{
-		Data: uintptr(ptr),
-		Len:  uintptr(size), // Tinygo requires these as uintptrs even if they are int fields.
-		Cap:  uintptr(size), // ^^ See https://github.com/tinygo-org/tinygo/issues/1284
-	}))
+	return unsafe.String((*byte)(unsafe.Pointer(uintptr(ptr))), size)
 }
 
 // stringToPtr returns a pointer and size pair for the given string in a way
 // compatible with WebAssembly numeric types.
+// The returned pointer aliases the string hence the string must be kept alive
+// until ptr is no longer needed.
 func stringToPtr(s string) (uint32, uint32) {
-	buf := []byte(s)
-	ptr := &buf[0]
-	unsafePtr := uintptr(unsafe.Pointer(ptr))
-	return uint32(unsafePtr), uint32(len(buf))
+	ptr := unsafe.Pointer(unsafe.StringData(s))
+	return uint32(uintptr(ptr)), uint32(len(s))
 }
 
 // stringToLeakedPtr returns a pointer and size pair for the given string in a way
-// compatible with WebAssembly numeric types. The pointer is not automatically
-// managed by TinyGo hence it must be freed by the host.
+// compatible with WebAssembly numeric types.
+// The pointer is not automatically managed by TinyGo hence it must be freed by the host.
 func stringToLeakedPtr(s string) (uint32, uint32) {
 	size := C.ulong(len(s))
```

Comment:
we aren't special casing len(s) == 0 here (to return ptr=0, size=0), and I don't recommend doing so as it is so complicated to explain. So fine as-is.

I'll explain anyway just to save the research on the special casing which was hard to track down.

Unlike `stringToPtr`, the host must call free on the ptr result of `stringToLeakedPtr`. If we returned ptr=0 manually, instead of proceeding to the logic below on len(s)==0, it would be ok even though it might actually slow the code.

When the host calls the "free" function exported by the guest with wasm i32(0), it ends up calling TinyGo's [`libc_free` function](https://github.com/tinygo-org/tinygo/blob/0d56dee00f49bd50eb373c02c30062a75ec28f10/src/runtime/arch_tinygowasm_malloc.go#LL25C6-L25C15).

TinyGo's compiler converts [zero to unsafe.Pointer(nil) on all architectures](https://github.com/tinygo-org/tinygo/blob/release/compiler/compiler.go#L2803), including wasm32, so this means `libc_free` is called with nil on i32(0), same as any other function export that has an `unsafe.Pointer` param. An unrelated side effect is, you can't represent memory offset 0 with an `unsafe.Pointer`, as it cannot be differentiated from nil.

Anyway, TinyGo's `libc_free` is following the [C specification for free](https://linux.die.net/man/3/free) which says "ptr is NULL, no operation is performed". This means the passing zero will work not just withthe default `libc_free`, but any other "free" export someone might have wired in.

Reply:
@ncruces mentioned that being unable to access offset 0 is a feature not a bug, and it could be a good practice to actually enforce code never tries to use offset zero.

If anyone ends up writing a reference based on these notes in TinyGo, our TinyGo site doc, tinymem RATIONALE.md, or just a personal site, this seems solid advice to consider.

Reply:
I should point out this quirk of `malloc`:

> If size is 0, then malloc() returns either NULL, or a unique pointer value that can later be successfully passed to free().

TinyGo will return nil for `malloc(0)` and I don't see a reason to change this, but there is the possibility that other heap implementations will do something different. (In fact, the internal TinyGo memory allocator `runtime.alloc` returns a non-nil pointer when allocating a zero-length object).

Also, unrelated:

> @ncruces mentioned that being unable to access offset 0 is a feature not a bug, and it could be a good practice to actually enforce code never tries to use offset zero.

There is actually a more subtle reason to enforce this:

  - Dereferencing a nil pointer will lead to a panic in TinyGo, as the Go language specification says it should.
  - If you work around this (using `unsafe` or whatever), you may find that LLVM _also_ assumes nil pointers are invalid. If it sees you are dereferencing a nil pointer, it will consider this undefined behavior and act accordingly (probably by deleting code as it assumes undefined behavior is unreachable).